File: common/src/main/java/com/jon/common/service/CotFactory.java
package com.jon.common.service;

import android.content.SharedPreferences;

import com.jon.common.cot.CursorOnTarget;
import com.jon.common.utils.Battery;

import java.util.List;

public abstract class CotFactory {
    protected final SharedPreferences prefs;
    protected final GpsCoords gpsCoords = GpsCoords.getInstance();
    protected final Battery battery = Battery.getInstance();

    protected CotFactory(SharedPreferences prefs) {
        this.prefs = prefs;
    }

    protected abstract List<CursorOnTarget> generate();
    protected abstract List<CursorOnTarget> initialise();
    protected abstract List<CursorOnTarget> update();

    protected abstract void clear();
}

File: common/src/main/java/com/jon/common/service/CotFactory.kt
package com.jon.common.service

import android.content.SharedPreferences
import com.jon.common.cot.CursorOnTarget
import com.jon.common.di.IBuildResources
import com.jon.common.repositories.IGpsRepository
import com.jon.common.repositories.IBatteryRepository
import com.jon.common.repositories.IDeviceUidRepository

abstract class CotFactory(
        protected val prefs: SharedPreferences,
        protected val buildResources: IBuildResources,
        protected val deviceUidRepository: IDeviceUidRepository,
        protected val gpsRepository: IGpsRepository,
        protected val batteryRepository: IBatteryRepository
) {
    abstract fun generate(): List<CursorOnTarget>
    protected abstract fun initialise(): List<CursorOnTarget>
    protected abstract fun update(): List<CursorOnTarget>
    abstract fun clear()
}
